### ISSUE TITLE
🎲 Phase 23.1 — UncertaintyQuantifier: Bayesian uncertainty estimation, confidence calibration & epistemic-aleatoric decomposition

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#530]
## Overview

**UncertaintyQuantifier** provides calibrated uncertainty estimates for all cognitive decisions. It decomposes uncertainty into **epistemic** (model/knowledge uncertainty — reducible with more data) and **aleatoric** (inherent randomness — irreducible) components, enabling the system to *know what it doesn't know*.

This is the foundation of Phase 23 — **Metacognitive Uncertainty & Confidence Calibration** — giving every downstream module a principled way to express, propagate, and act on uncertainty rather than treating point estimates as ground truth.

### Key Capabilities
- **Ensemble-based uncertainty** via prediction disagreement (variance, mutual information)
- **Epistemic–aleatoric decomposition** for targeted uncertainty reduction
- **Post-hoc calibration** (Platt scaling, isotonic regression, temperature scaling, beta calibration)
- **Background recalibration** loop for drift-aware confidence maintenance
- **Shannon entropy** computation for information-theoretic uncertainty characterisation

### Integration Points
| Component | Relationship |
|---|---|
| `CreativeOrchestrator` (22.5) | Confidence gates on generated concepts |
| `AestheticEvaluator` (22.4) | Calibrated scores for novelty/value judgements |
| `ReasoningOrchestrator` (20.5) | Uncertainty-aware inference chains |
| `AffectiveOrchestrator` (21.5) | Confidence on emotion classifications |
| `ReflectionCycle` (16.5) | Epistemic uncertainty triggers self-improvement |
| `SurpriseDetector` (13.4) | High epistemic uncertainty ↔ surprise signal |

---

### `UncertaintyType` — enum (str, Enum)

```python
class UncertaintyType(str, Enum):
    EPISTEMIC = "epistemic"      # model/knowledge uncertainty — reducible
    ALEATORIC = "aleatoric"      # inherent randomness — irreducible
    MIXED     = "mixed"          # both components significant
```

### `CalibrationMethod` — enum (str, Enum)

```python
class CalibrationMethod(str, Enum):
    PLATT_SCALING         = "platt_scaling"          # logistic regression on logits
    ISOTONIC_REGRESSION   = "isotonic_regression"    # non-parametric monotonic fit
    TEMPERATURE_SCALING   = "temperature_scaling"    # single scalar temperature T
    BETA_CALIBRATION      = "beta_calibration"       # beta distribution fit (Kull et al. 2017)
```

### `UncertaintyEstimate` — frozen dataclass

```python
@dataclass(frozen=True, slots=True)
class UncertaintyEstimate:
    point_estimate: float                       # best single-value prediction
    confidence_interval: tuple[float, float]    # (lower, upper) at configured level
    uncertainty_type: UncertaintyType           # dominant uncertainty source
    epistemic_component: float                  # ∈ [0, 1] — knowledge uncertainty
    aleatoric_component: float                  # ∈ [0, 1] — inherent randomness
    calibrated: bool                            # True if post-hoc calibration applied
    entropy: float                              # Shannon entropy in bits (≥ 0)
    created_at: datetime                        # UTC timestamp

    def __post_init__(self) -> None:
        if not (0.0 <= self.epistemic_component <= 1.0):
            raise ValueError("epistemic_component must be in [0, 1]")
        if not (0.0 <= self.aleatoric_component <= 1.0):
            raise ValueError("aleatoric_component must be in [0, 1]")
        if self.confidence_interval[0] > self.confidence_interval[1]:
            raise ValueError("lower bound must ≤ upper bound")
        if self.entropy < 0.0:
            raise ValueError("entropy must be non-negative")
```

### `EnsembleDisagreement` — frozen dataclass

```python
@dataclass(frozen=True, slots=True)
class EnsembleDisagreement:
    predictions: tuple[float, ...]    # individual ensemble member outputs
    mean: float                       # arithmetic mean
    variance: float                   # sample variance (disagreement proxy)
    mutual_information: float         # I(y; θ|x) — epistemic uncertainty proxy

    def __post_init__(self) -> None:
        if len(self.predictions) < 1:
            raise ValueError("need at least one prediction")
        if self.variance < 0.0:
            raise ValueError("variance must be non-negative")
        if self.mutual_information < 0.0:
            raise ValueError("mutual_information must be non-negative")
```

### `QuantifierConfig` — frozen dataclass

```python
@dataclass(frozen=True, slots=True)
class QuantifierConfig:
    n_ensemble_members: int = 5                                      # MC-dropout or deep ensemble size
    calibration_method: CalibrationMethod = CalibrationMethod.PLATT_SCALING
    confidence_level: float = 0.95                                   # CI coverage probability
    epistemic_threshold: float = 0.3                                 # ratio above which type=EPISTEMIC
    recalibration_interval_s: float = 3600.0                         # background recalibration period

    def __post_init__(self) -> None:
        if self.n_ensemble_members < 1:
            raise ValueError("need at least 1 ensemble member")
        if not (0.0 < self.confidence_level < 1.0):
            raise ValueError("confidence_level must be in (0, 1)")
        if not (0.0 <= self.epistemic_threshold <= 1.0):
            raise ValueError("epistemic_threshold must be in [0, 1]")
```

---

## Protocol

```python
from typing import Any, Protocol, Sequence, runtime_checkable

@runtime_checkable
class UncertaintyQuantifier(Protocol):
    """Bayesian uncertainty estimation with epistemic-aleatoric decomposition."""

    async def quantify(
        self,
        predictions: Sequence[float],
        context: dict[str, Any] | None = None,
    ) -> UncertaintyEstimate:
        """Produce a calibrated uncertainty estimate from ensemble predictions."""
        ...

    async def calibrate(
        self,
        predictions: Sequence[float],
        actuals: Sequence[float],
    ) -> None:
        """Fit calibration parameters from (predicted, actual) pairs."""
        ...

    async def decompose(
        self,
        prediction: float,
        ensemble: EnsembleDisagreement,
    ) -> UncertaintyEstimate:
        """Decompose total uncertainty into epistemic and aleatoric components."""
        ...

    def is_calibrated(self) -> bool:
        """Return True if calibration parameters have been fitted."""
        ...
```

---

## Implementation: `BayesianUncertaintyQuantifier`

```python
class BayesianUncertaintyQuantifier:
    """Production implementation — ensemble disagreement + post-hoc calibration."""

    def __init__(self, config: QuantifierConfig | None = None) -> None:
        self._config = config or QuantifierConfig()
        self._lock = asyncio.Lock()              # guards calibration state
        self._calibration_params: dict[str, float] | None = None
        self._calibration_data: list[tuple[float, float]] = []
        self._recal_task: asyncio.Task[None] | None = None

    # ── quantify ────────────────────────────────────────────────
    async def quantify(
        self,
        predictions: Sequence[float],
        context: dict[str, Any] | None = None,
    ) -> UncertaintyEstimate:
        """
        1. Build EnsembleDisagreement from predictions
        2. Compute mutual information (epistemic proxy)
        3. Compute data variance (aleatoric proxy)
        4. Apply calibration if fitted
        5. Classify UncertaintyType via epistemic_threshold
        6. Build confidence interval from quantiles
        7. Increment asi_uncertainty_quantifications_total
        """
        ...

    # ── calibrate ───────────────────────────────────────────────
    async def calibrate(
        self,
        predictions: Sequence[float],
        actuals: Sequence[float],
    ) -> None:
        """
        Fit calibration curve under _lock.
        - PLATT_SCALING: logistic regression on logits → σ(a·z + b)
        - ISOTONIC_REGRESSION: monotonic step function via PAVA
        - TEMPERATURE_SCALING: minimise NLL w.r.t. scalar T
        - BETA_CALIBRATION: fit Beta(a, b, c) (Kull et al.)
        Update asi_uncertainty_calibration_error gauge (ECE).
        Increment asi_uncertainty_recalibrations_total.
        """
        ...

    # ── decompose ───────────────────────────────────────────────
    async def decompose(
        self,
        prediction: float,
        ensemble: EnsembleDisagreement,
    ) -> UncertaintyEstimate:
        """
        Decompose total variance:
          total_var     = Var[y|x]
          epistemic_var = Var_θ[E[y|x,θ]]    ≈ ensemble.variance
          aleatoric_var = E_θ[Var[y|x,θ]]    ≈ total_var - epistemic_var
        Normalise to [0, 1] and classify via epistemic_threshold.
        Record asi_uncertainty_epistemic_ratio histogram.
        """
        ...

    # ── is_calibrated ───────────────────────────────────────────
    def is_calibrated(self) -> bool:
        return self._calibration_params is not None

    # ── background recalibration ────────────────────────────────
    async def start(self) -> None:
        """Start background recalibration loop."""
        self._recal_task = asyncio.create_task(self._recalibration_loop())

    async def stop(self) -> None:
        """Cancel recalibration loop and await cleanup."""
        if self._recal_task:
            self._recal_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._recal_task

    async def _recalibration_loop(self) -> None:
        """Every recalibration_interval_s: re-fit calibration from accumulated data."""
        while True:
            await asyncio.sleep(self._config.recalibration_interval_s)
            if len(self._calibration_data) > 0:
                preds, acts = zip(*self._calibration_data)
                await self.calibrate(list(preds), list(acts))
```

### Calibration Algorithm Detail

**Platt Scaling** (default):
```
P(y=1|z) = σ(a·z + b)
```
Fit `a`, `b` via maximum likelihood on validation set. Single forward pass at inference.

**Expected Calibration Error (ECE)**:
```
ECE = Σ_b (|B_b|/N) · |acc(B_b) - conf(B_b)|
```
Partitions predictions into `B` equal-width bins; measures mean absolute gap between accuracy and confidence.

### Entropy Computation

```
H(p) = -Σ p_i · log₂(p_i)
```
For continuous predictions, discretise into histogram bins or use differential entropy estimate.

---

## `NullUncertaintyQuantifier`

```python
class NullUncertaintyQuantifier:
    """No-op implementation for testing and dependency injection."""

    async def quantify(self, predictions, context=None):
        return UncertaintyEstimate(
            point_estimate=float(sum(predictions) / len(predictions)),
            confidence_interval=(0.0, 1.0),
            uncertainty_type=UncertaintyType.MIXED,
            epistemic_component=0.5,
            aleatoric_component=0.5,
            calibrated=False,
            entropy=0.0,
            created_at=datetime.now(UTC),
        )

    async def calibrate(self, predictions, actuals):
        pass  # no-op

    async def decompose(self, prediction, ensemble):
        return await self.quantify(ensemble.predictions)

    def is_calibrated(self) -> bool:
        return False
```

---

## Factory

```python
def make_uncertainty_quantifier(
    config: QuantifierConfig | None = None,
) -> UncertaintyQuantifier:
    """Create the default BayesianUncertaintyQuantifier."""
    return BayesianUncertaintyQuantifier(config)
```

---

## Prometheus Metrics

| # | Name | Type | Description |
|---|------|------|-------------|
| 1 | `asi_uncertainty_quantifications_total` | Counter | Total `quantify()` calls |
| 2 | `asi_uncertainty_calibration_error` | Gauge | Expected Calibration Error (ECE) |
| 3 | `asi_uncertainty_epistemic_ratio` | Histogram | epistemic / (epistemic + aleatoric) distribution |
| 4 | `asi_uncertainty_entropy_bits` | Histogram | Shannon entropy of prediction sets |
| 5 | `asi_uncertainty_recalibrations_total` | Counter | Background recalibration events |

### PromQL Examples

```promql
# Quantification rate (per minute)
rate(asi_uncertainty_quantifications_total[5m]) * 60

# Current calibration quality
asi_uncertainty_calibration_error

# Mean epistemic ratio (high = model needs more data)
histogram_quantile(0.5, rate(asi_uncertainty_epistemic_ratio_bucket[5m]))

# Recalibration frequency
rate(asi_uncertainty_recalibrations_total[1h]) * 3600
```

### Grafana Alert YAML

```yaml
- alert: HighCalibrationError
  expr: asi_uncertainty_calibration_error > 0.15
  for: 10m
  labels:
    severity: warning
  annotations:
    summary: "ECE > 15% — predictions are poorly calibrated"

- alert: HighEpistemicUncertainty
  expr: histogram_quantile(0.9, rate(asi_uncertainty_epistemic_ratio_bucket[5m])) > 0.7
  for: 15m
  labels:
    severity: warning
  annotations:
    summary: "90th percentile epistemic ratio > 0.7 — model may need retraining"

- alert: RecalibrationStalled
  expr: rate(asi_uncertainty_recalibrations_total[2h]) == 0
  for: 2h
  labels:
    severity: warning
  annotations:
    summary: "No recalibration in 2h — check background loop"
```

---

## `mypy` Narrowing Table

| Pattern | Narrowed Type | Notes |
|---|---|---|
| `match estimate.uncertainty_type:` | | |
| `  case UncertaintyType.EPISTEMIC` | `Literal[UncertaintyType.EPISTEMIC]` | epistemic_component > threshold |
| `  case UncertaintyType.ALEATORIC` | `Literal[UncertaintyType.ALEATORIC]` | aleatoric dominates |
| `  case UncertaintyType.MIXED` | `Literal[UncertaintyType.MIXED]` | both significant |
| `match config.calibration_method:` | | |
| `  case CalibrationMethod.PLATT_SCALING` | `Literal[CalibrationMethod.PLATT_SCALING]` | logistic fit |
| `  case CalibrationMethod.ISOTONIC_REGRESSION` | `Literal[CalibrationMethod.ISOTONIC_REGRESSION]` | PAVA |
| `  case CalibrationMethod.TEMPERATURE_SCALING` | `Literal[CalibrationMethod.TEMPERATURE_SCALING]` | scalar T |
| `  case CalibrationMethod.BETA_CALIBRATION` | `Literal[CalibrationMethod.BETA_CALIBRATION]` | Beta(a,b,c) |
| `isinstance(q, UncertaintyQuantifier)` | `UncertaintyQuantifier` | runtime_checkable Protocol |

---

## Test Targets (12)

| # | Test | Assertion |
|---|------|-----------|
| 1 | `test_protocol_compliance` | `isinstance(BayesianUncertaintyQuantifier(), UncertaintyQuantifier)` |
| 2 | `test_ensemble_disagreement_variance` | high variance → wide confidence interval |
| 3 | `test_calibration_reduces_ece` | ECE decreases after `calibrate()` |
| 4 | `test_epistemic_decomposition` | ensemble variance → epistemic component |
| 5 | `test_aleatoric_decomposition` | data noise → aleatoric component |
| 6 | `test_mixed_uncertainty_type` | both components above threshold → MIXED |
| 7 | `test_config_validation` | invalid confidence_level raises ValueError |
| 8 | `test_null_impl_returns_defaults` | NullUncertaintyQuantifier always returns uncalibrated |
| 9 | `test_factory_returns_protocol` | `make_uncertainty_quantifier()` returns UncertaintyQuantifier |
| 10 | `test_recalibration_loop` | background task triggers recalibration |
| 11 | `test_single_prediction_edge_case` | single value → zero variance, ALEATORIC |
| 12 | `test_nan_handling` | NaN in predictions raises or is handled gracefully |

### Test Skeleton — `test_calibration_reduces_ece`

```python
@pytest.mark.asyncio
async def test_calibration_reduces_ece() -> None:
    """After calibrate(), ECE should decrease."""
    q = BayesianUncertaintyQuantifier(QuantifierConfig(
        calibration_method=CalibrationMethod.PLATT_SCALING,
    ))
    # Overconfident predictions: all 0.9, but only 60% correct
    preds = [0.9] * 100
    actuals = [1.0] * 60 + [0.0] * 40

    est_before = await q.quantify(preds)
    await q.calibrate(preds, actuals)
    est_after = await q.quantify(preds)

    assert q.is_calibrated()
    assert est_after.calibrated is True
    # After calibration, point_estimate should be closer to 0.6
    assert abs(est_after.point_estimate - 0.6) < abs(est_before.point_estimate - 0.6)
```

### Test Skeleton — `test_epistemic_decomposition`

```python
@pytest.mark.asyncio
async def test_epistemic_decomposition() -> None:
    """High ensemble disagreement → high epistemic component."""
    q = BayesianUncertaintyQuantifier(QuantifierConfig(
        epistemic_threshold=0.3,
    ))
    # Ensemble members wildly disagree
    ensemble = EnsembleDisagreement(
        predictions=(0.1, 0.9, 0.3, 0.8, 0.2),
        mean=0.46,
        variance=0.122,
        mutual_information=0.85,
    )
    est = await q.decompose(0.46, ensemble)

    assert est.uncertainty_type == UncertaintyType.EPISTEMIC
    assert est.epistemic_component > est.aleatoric_component
    assert est.epistemic_component > 0.3
```

---

## Implementation Order

| Step | Artefact | LoC |
|------|----------|-----|
| 1 | `UncertaintyType` enum | ~8 |
| 2 | `CalibrationMethod` enum | ~10 |
| 3 | `UncertaintyEstimate` dataclass + `__post_init__` | ~30 |
| 4 | `EnsembleDisagreement` dataclass + computed props | ~20 |
| 5 | `QuantifierConfig` dataclass + defaults + validation | ~20 |
| 6 | `UncertaintyQuantifier` Protocol | ~15 |
| 7 | `BayesianUncertaintyQuantifier` skeleton (`__init__`) | ~25 |
| 8 | `quantify()` — ensemble disagreement + CI | ~45 |
| 9 | `decompose()` — epistemic/aleatoric split | ~35 |
| 10 | `calibrate()` — Platt/isotonic/temperature/beta dispatch | ~60 |
| 11 | Background `_recalibration_loop` + `start()`/`stop()` | ~30 |
| 12 | `NullUncertaintyQuantifier` | ~20 |
| 13 | `make_uncertainty_quantifier()` factory | ~8 |
| 14 | Prometheus metrics (5 instruments) | ~25 |
| | **Total** | **~351** |

---

## Phase 23 Sub-Phase Tracker

| Sub-Phase | Component | Issue | Status |
|-----------|-----------|-------|--------|
| 23.1 | UncertaintyQuantifier | **this issue** | 🔵 Spec'd |
| 23.2 | ConfidenceCalibrator | TBD | ⬜ Pending |
| 23.3 | BeliefUpdater | TBD | ⬜ Pending |
| 23.4 | RiskEstimator | TBD | ⬜ Pending |
| 23.5 | MetacognitiveOrchestrator | TBD | ⬜ Pending |
